### ISSUE TITLE
[ast] NewPulseWhenSrcBusy assertion error

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#13893]
This assertion error happens when it's switching to use the external clock.

> "../src/lowrisc_systems_ast_0.1/rtl/ast_pulse_sync.sv", 144: tb.dut.u_ast.u_rng.u_rng_val_pulse_sync.NewPulseWhenSrcBusy: started at 3133040170ps failed at 3133040170ps
> 	Offending '(!src_busy_o)'
> UVM_ERROR @ 3133.040170 us: (ast_pulse_sync.sv:144) [ASSERT FAILED] NewPulseWhenSrcBusy

Looks like one of the clocks is paused during the clock switching and another RNG req may come in before the previous one is completed. Thanks @tjaychen for debugging this. Please add more details for this issue.

To reproduce this issue, get the PR #13892 and run this chip-level test
> util/dvsim/dvsim.py  hw/top_earlgrey/dv/chip_sim_cfg.hjson -s 1444919156 -r 1

**Comment by tjaychen** [comment]
i should help add some explanation here. 
It seems what happens in this test is that the rng inside ast is enabled.

When software tries to make an external clock switch, one of the clocks on the `ast_pulse_sync` is temporarily stopped due to the glitchless switch process.  However, the clock on the other side is still going, so it violates the assertion since one side basically stops responding. '

@Jacob-Levy / @arnonsha / @zi-v it would be good if you guys can give some guidance on how this should behave. My perspective is that the switch should not impact the rng behavior, so we either need to tweak the assertion or the logic (assuming it's the real one) to handle the case where one side is temporarily without clocks.

**Comment by arnonsha** [comment]
@Jacob-Levy can you please take a look ?

**Comment by Jacob-Levy** [comment]
@tjaychen  There were two (2) issues in the prim_pulse_sync.sv cell that were fixed by the ast_pulse_sync.sv
   1) The resets from both side needs to reset the flops in the cell.
   2) The prim_pulse_sync.sv dose not implement 'busy' when hadling  a pulse.  moreover a new pulse is not blocked 
       when the current pulse is still working (no assert too).

I suggest that I will open an issue on the prim_pulse_sync.sv that describe both issue and suggest that 
the ast_pulse_sync.sv code will be used. There are few mails between @Jacob-Levy  and @tjaychen thatdescribe 
the fails of the cell.

As for the specific fall, why a new pulse get introduced before the previous pulse fininshed?

**Comment by Jacob-Levy** [comment]
You can bypass the issue by stopping the RNG when you switching to external clock. Once the switch is done , you can reenable the RNG. The RNG clock is the system clock that is affected by the switch to external clock.
In real silicon the new pulse will be ignored (blocked) and the "alert" not fire...

We can reduce the allert to warning if needed...

**Comment by Jacob-Levy** [comment]
In the prim_pulse_sync.sv the check/ASSERT is conditioed by INC_ASSERT.
```
  // source active must come far enough such that the destination domain has time
  // to create a valid pulse.
`ifdef INC_ASSERT
  logic src_active_flag;

  // source active flag tracks whether there is an ongoing "toggle" event.
  // Until this toggle event is accepted by the destination domain (negative edge of
  // of the pulse output), the source side cannot toggle again.
  always_ff @(posedge clk_src_i or negedge dst_pulse_o or negedge rst_src_ni) begin
    if (!rst_src_ni) begin
      src_active_flag <= '0;
    end else if (!dst_pulse_o && src_active_flag) begin
      src_active_flag <= '0;
    end else if (src_pulse_i) begin
      src_active_flag <= 1'b1;
    end
  end

 `ASSERT(SrcPulseCheck_M, src_pulse_i |-> !src_active_flag, clk_src_i, !rst_src_ni)
`endif
```

A can add the same condition in the ast_pulse_sync.sv
```
`ifdef INC_ASSERT
// A new PULSE can only be introduced when source is not BUSY.
`ASSERT(NewPulseWhenSrcBusy, $rose(src_pulse_i) |-> !src_busy_o, clk_src_i, !rst_src_n)
`endif
 ```

**Comment by Jacob-Levy** [comment]
After looking at the rng.sv code, the issue you are seeing might releate to the 'srate_value..
The clocks involve is RCLK (TLUL) (24MHz) and RNG (SYS) (100MHz), both are affected during external switching.

The 'srate_value' is updated by RCLK and if the 'dv_srate_value' is set too low, it might generate a new pulse while the
perivous one is still active. I will check if I can use the 'src_busy' to block/hold the new pulse until the previous one clears.

@weicaiyang It whould be helpfull to see the waves so I'll make sure I fully understand the issue.

**Comment by Jacob-Levy** [comment]
I can fix the issue in the rmg.sv by holding the generation of the `srate_rng_val` pulse when `src_busy` is active.
```
logic src_busy;

always_ff @( posedge clk_i, negedge rst_n ) begin
  if ( !rst_n ) begin
    srate_cnt     <= 12'h000;
    srate_rng_val <= 1'b0;
  end else if ( (srate_cnt == srate_value) && src_busy ) begin
    srate_rng_val <= 1'b0;
  end else if ( srate_cnt == srate_value ) begin
    srate_cnt     <= 12'h000;
    srate_rng_val <= 1'b1;
  end else begin
    srate_cnt     <= srate_cnt + 1'b1;
    srate_rng_val <= 1'b0;
  end
end
```

**Comment by Jacob-Levy** [comment]
@weicaiyang With this fix, I can leave all the assertion in place. 
Attached id the modified rng.sv. Can you test the fix? If it works I will PR it.

**Comment by tjaychen** [comment]
@Jacob-Levy yes i think your description of `srate_value` is what i remember as well. I think that count doesn't stop, and the eventual count triggers another pulse is what i roughly remember. 

and yes you're right the current `pulse_sync` does not handle the extra pulse before sync is done, so we added an assertion. I think our thinking at the time was that even if we added that busy, this may not have been intended behavior from the source. So it's better to force the failure, and let the source side deal with the extra pulse in whatever way it deems best.

regarding the fix @Jacob-Levy , why not just push it as a PR first? we can pull it down to try.

**Comment by Jacob-Levy** [comment]
@tjaychen  I wanted him to use the failing case as a qualifier to my fix by simply replacing the rng.sv. The fix is very local and simple and only suspend the ofending pulse.

**Comment by Jacob-Levy** [comment]
@tjaychen BTW, the prim_pulse_sync.sv does not block the extra pulse like in the ast_pulse_sync.sv. The assertion code, if enabled, creats an only an error. After synthesis, the ASSERT is gone and the next pulse is not blocked.

**Comment by weicaiyang** [comment]
I haven't run the test yet, but just look at the change. `src_busy` isn't not assigned with any value. ` if ( (srate_cnt == srate_value) && src_busy )` is always false. Seems like the change won't make any difference, or I may miss something.


```
logic src_busy;

always_ff @( posedge clk_i, negedge rst_n ) begin
  if ( !rst_n ) begin
    srate_cnt     <= 12'h000;
    srate_rng_val <= 1'b0;
  end else if ( (srate_cnt == srate_value) && src_busy ) begin
    srate_rng_val <= 1'b0;
  end else if ( srate_cnt == srate_value ) begin
    srate_cnt     <= 12'h000;
    srate_rng_val <= 1'b1;
  end else begin
    srate_cnt     <= srate_cnt + 1'b1;
    srate_rng_val <= 1'b0;
  end
end
```

**Comment by tjaychen** [comment]
> @tjaychen BTW, the prim_pulse_sync.sv does not block the extra pulse like in the ast_pulse_sync.sv. The assertion code, if enabled, creats an only an error. After synthesis, the ASSERT is gone and the next pulse is not blocked.

yes that's right. Our intent is to flag them at the design state and force the design to deal with the corner case. If it is allowed (ie new pulse when handshake is ongoing), then the design should swallow the pulse. If it is not allowed, then the design should adjust accordingly.

**Comment by Jacob-Levy** [comment]
It is the output of the ast_pulse_sync cell instatiated just bellow.

**Comment by tjaychen** [comment]
> > @tjaychen BTW, the prim_pulse_sync.sv does not block the extra pulse like in the ast_pulse_sync.sv. The assertion code, if enabled, creats an only an error. After synthesis, the ASSERT is gone and the next pulse is not blocked.
> 
> yes that's right. Our intent is to flag them at the design state and force the design to deal with the corner case. If it is allowed (ie new pulse when handshake is ongoing), then the design should swallow the pulse. If it is not allowed, then the design should adjust accordingly.

having said that @Jacob-Levy, why don't you raise this point as a separate issue? and let's bring it to the biweekly silicon meeting for more community input. It might be that people prefer a busy approach natively.  Those meetings are usually good for discussions like that.

**Comment by Jacob-Levy** [comment]
I will open an issue on prim_pulse_sync.sv cell with the two issues I mentioned and put all the information in the issue.

**Comment by weicaiyang** [comment]
> It is the output of the ast_pulse_sync cell instatiated just bellow.
> 
> ```
> ////////////////////////////////////////
> // Sychronize Bus & Valid to RNG Clock
> ////////////////////////////////////////
> logic sync_rng_val, srate_rng_val_en;
> 
> ast_pulse_sync u_rng_val_pulse_sync (
>   .scan_mode_i ( scan_mode_i ),
>   .scan_reset_ni ( rst_n ),
>   // source clock domain
>   .clk_src_i ( clk_i ),
>   .rst_src_ni ( rst_n ),
>   .src_pulse_i ( srate_rng_val ),
>   .src_pulse_en_o ( srate_rng_val_en ),
>   .src_busy_o ( src_busy ),
>   // destination clock domain
>   .clk_dst_i ( clk_ast_rng_i ),
>   .rst_dst_ni ( rst_ast_rng_ni ),
>   .dst_pulse_o ( sync_rng_val )
> );
> ```

oh, I see. `src_busy` isn't a new variable. You just move the declaration.

**Comment by Jacob-Levy** [comment]
See PR [AST] Fix RNG srate_rng_val and update comments #14280
